man/dg535.1.in
.TH dg535 1
.SH NAME
dg535 \- control SRS DG535 Digital Delay/Pulse Generator
.SH SYNOPSIS
.nf
.B dg535 [\fIOPTIONS\fR]

.fi
.SH DESCRIPTION
\fBdg535\fR controls the Stanford Research Systems, Inc. model DG535 
four channel digital delay/pulse generator via its GPIB interface.
\fBdg535\fR performs actions in the order they appear on the command line.
.SH "OPTIONS, GENERAL"
\fBdg535\fR accepts the following command line options:
.TP
\fB\-n\fR, \fB\-\-name\fR name
Set the name used to look up the instrument address in
the config file.  The default is \fIdg535\fR.
.TP
\fB\-a\fR, \fB\-\-address\fR address
Set the instrument address.
See gpib-utils.conf(5) for a description of address formats.
.TP
\fB\-c\fR, \fB\-\-clear\fR
Clear the instrument and load default configuration:
.IP
\fItrigger mode:\fR single-shot, i.e. triggers are off
.br
\fIinternal:\fR 10kHz
.br
\fIburst mode:\fR 10kHz, 10 pulses per burst, 20 periods per burst
.br
\fIexternal:\fR +1 VDC, positive slope, high-z termination
.br
\fIdelays:\fR All delays (A,B,C,D) are linked to T0 and set to zero.
.br
\fIoutputs:\fR All outputs set to drive high-z loads to TTL levels.
.TP
\fB\-e\fR, \fB\-\-display\-string\fR "string"
Display string on the front panel LCD.  Strings greater than 20 characters
will be truncated.  
Spaces are stripped so use the underline character to separate words.
An empty string clears the display.
.TP
\fB\-l\fR, \fB\-\-local\fR
Enable the front panel.
.SH "OPTIONS, OUTPUT"
.TP
\fB\-N\fR, \fB\-\-out\-chan\fR t0|a|b|ab|c|d|cd
Select an output channel for one of the output configuration commands.
.TP
\fB\-d\fR, \fB\-\-get\-delay\fR
Get delay configuration for channel specified with \-\-out\-chan.
Delay is expressed in seconds.
.TP
\fB\-D\fR, \fB\-\-set\-delay\fR chan,secs
Set delay configuration for channel specified with \-\-out\-chan.
The delay is relative to the specified channel, and must include
units: s, ms, us, ns, ps).
.TP
\fB\-q\fR, \fB\-\-get\-out\-mode\fR
Get output mode for channel specified with \-\-out\-chan.
.TP
\fB\-Q\fR, \fB\-\-set\-out\-mode\fR ttl|nim|ecl|var
Set output mode for channel specified with \-\-out\-chan.
If \fIttl\fR, \fInim\fR, or \fIecl\fR is selected, the polarity is selected
with the \-\-set-out-polarity options.
If \fIvar\fR is selected, the output amplitude and offsets may be
set with the \-\-set\-out\-ampl and \-\-set\-out\-offset options.
.TP
\fB\-g\fR, \fB\-\-get\-out\-ampl\fR
Get output amplitude in volts for channel specified with \-\-out\-chan.
.TP
\fB\-G\fR, \fB\-\-set\-out\-ampl\fR ampl
Set output amplitude in volts for channel specified with --out-chan
if the output is in variable mode.  The maximum amplitude plus offset
is limited to plus or minus 4VDC, while the minimum is limited to 
plus or minus 0.1VDC.
.TP
\fB\-f\fR, \fB\-\-get-out-offset\fR
Get output offset in volts for channel specified with \-\-out\-chan.
.TP
\fB\-F\fR, \fB\-\-set\-out\-offset\fR offset
Set output offset in volts for channel specified with \-\-out\-chan
if the output is in variable mode.
For pulses which have a rising edge output, the offset is the voltage 
between the lowest output level and ground.
So a TTL signal has zero offset and +4VDC amplitude, while
an ECL signal has -1.8VDC offset and +1VDC amplitude.
To invert the polarity of the pulse, i.e. for a falling edge output,
a TTL output is described as a +4VDC offset with a 4VDC amplitude
pulse, while an ECL output is described as a -0.8VDC offset with
a 1.0VDC amplitude.
.TP
\fB\-p\fR, \fB\-\-get-out-polarity\fR
Get output polarity for channel specified with \-\-out\-chan.
.TP
\fB\-P\fR, \fB\-\-set\-out\-polarity\fR +|-
Set output polarity for channel specified with \-\-out\-chan.
.TP
\fB\-y\fR, \fB\-\-get\-out\-z\fR
Get output impedence for channel specified with \-\-out\-chan.
.TP
\fB\-Y\fR, \fB\-\-set\-out\-z\fR hi|lo
Set termination impedence for channel specified with \-\-out\-chan
to \fIlo\fR to drive a 50 ohm load, or \fIhi\fR to drive a high
impedence load.
.SH "OPTIONS, TRIGGER"
.TP
\fB\-x\fR, \fB\-\-single\-shot\fR
Trigger the instrument one time.  The instrument must be in single shot
trigger mode.
.TP
\fB\-m\fR, \fB\-\-get\-trig\-mode\fR
Get trigger mode.
.TP
\fB\-M\fR, \fB\-\-set\-trig\-mode\fR int|ext|ss|bur
Set trigger mode to \fIinternal\fR (int), \fIexternal\fR (ext),
\fIsingle-shot\fR (ss), or \fIburst\fR (bur).
.TP
\fB\-t\fR, \fB\-\-get\-trig\-rate\fR
Get the trigger rate.
.TP
\fB\-t\fR, \fB\-\-set\-trig\-rate\fR freq
Set the trigger rate to a aule in the range of 0.001Hz to 1.000MHz.
The precision is 0.001Hz below 10Hz, or 4 digits if above 10Hz.
The internal rate generator will settle to the programmed value within
2 seconds after the command is executed.  Significant departures may
be expected from the programmed value during the 2 second settling period.
Depending on the application, it may be useful to select single-shot trigger
mode during the settling time.
.TP
\fB\-s\fR, \fB\-\-get\-trig\-slope\fR
Get the trigger slope.
.TP
\fB\-S\fR, \fB\-\-set\-trig\-slope\fR rising|falling
Set the trigger slope condition for external triggers.
.TP
\fB\-b\fR, \fB\-\-get\-burst\-count\fR
Get burst count.
.TP
\fB\-B\fR, \fB\-\-set\-burst\-count\fR count
Set the number of pulses (2-32766) which will be in each burst of pulses 
when in the burst mode.
.TP
\fB\-z\fR, \fB\-\-get\-trig\-z\fR
Get impedence of external trigger input.
.TP
\fB\-Z\fR, \fB\-\-set\-trig\-z\fR hi|lo
Set impedence of external trigger input to \fIlo\fR (50 ohm)
or \fIhi\fR (1 Mohm).
.SH FILES
@X_SYSCONFDIR@/gpib-utils.conf
.br
~/.gpib-utils.conf
.SH SEE ALSO
gpib-utils.conf(5)
.PP
"Model DG535 Digital Delay / Pulse Generator Operation and Service Manual", 
revision 2.6, Nov. 2000.


NEWS
=========================================================================
gpib-utils-1.5 Release Notes                                   2014-08-21
=========================================================================

Moving code from sourceforge to github:

* Stop tracking autotools build products, etc.

* Remove ChangeLog (use commit log and NEWS), add README.md

These changes were staged in 2011:

* Export libvxi11.so library for general use.

* libvxi11 API cleanup, portability fix to gpib.c [Alessio Bogani].

* New utilities
  - ibquery:  Generic query utility.

=========================================================================
gpib-utils-1.4 Release Notes                                   2009-02-02 
=========================================================================

* Added support for new devices:
  - ics8064:  ICS Electronics 8064 Ethernet to Relay interface
  - a34401:   Agilent 34401A/34410A/34411A DMM.
  - hp6038:   Hewlett-Packard 603xA System DC Power Supply
  - hp6626:   Hewlett-Packard 6625A/6626A/6628A/6629A System DC Power Supply
  - hp6632:   Hewlett-Packard 6632A/6633A/6634A System DC Power Supply

* Handle non-zero GPIB board index and secondary GPIB addresses.

* Changed ics8065 utility to more general purpose icsconfig utility.

* Added automake/autoconf infrastructure so gpib-utils compiles with and
  without linux-gpib.

* Refactored ICS, VXI-11, and LSD code into (non-installed) libraries.

=========================================================================
gpib-utils-1.3 Release Notes                                   2007-11-23 
=========================================================================

* VXI-11 support for devices behind a Ethernet-GPIB gateway such as
  ICS 8065, or with built-in Ethernet ports such as the Agilent 6000 
  series oscilloscopes.

* Added config file: /etc/gpib-utils.conf.  Place device addresses here
  and overridde with -a option on command line.  If linux-gpib is used,
  /etc/gpib.conf need only contain bus master device information.

* Added support for new devices:
  - a6032 Agilent 6000-series oscilloscopes
  - dg535 Stanford Research Systems DG535 digital delay/pulse generator 
  - sr620 Stanford Research Systems SR620 Universal Time Interval Counter
  - ics8065 ICS Electroncs 8065 Ethernet-GPIB Controller
  - hp3488 Hewlett-Packard 3488A Switch/Control Unit

* Removed support for old instruments that I no longer have access to
  and which I know of nobody using:
  - r5005 Racal Dana model 5005 DMM
  - hp3455 Hewlett-Packard 3455A DMM
  - tek2230 Tektronix 2230 oscilloscope (never really working)


src/dg535.c
/* This file is part of gpib-utils.
   For details, see http://sourceforge.net/projects/gpib-utils.

   Copyright (C) 2005-2009 Jim Garlick <garlick.jim@gmail.com>

   gpib-utils is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   gpib-utils is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with gpib-utils; if not, write to the Free Software Foundation, 
   Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA */

/* Stanford Research Systems DG535 Digital Delay / Pulse Generator.
 */

#if HAVE_CONFIG_H
#include "config.h"
#endif
#define _GNU_SOURCE /* for HUGE_VALF */
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <errno.h>
#include <string.h>
#include <unistd.h>
#include <libgen.h>
#if HAVE_GETOPT_LONG
#include <getopt.h>
#endif
#include <assert.h>
#include <sys/time.h>
#include <math.h>
#include <stdint.h>

#include "libutil/util.h"
#include "gpib.h"

#define INSTRUMENT "dg535"

#define DG535_DO_TRIG                   0       /* trigger input */
#define DG535_DO_T0                     1       /* T0 output */
#define DG535_DO_A                      2       /* A output */
#define DG535_DO_B                      3       /* B output */
#define DG535_DO_AB                     4       /* AB and -AB outputs */
#define DG535_DO_C                      5       /* C output */
#define DG535_DO_D                      6       /* D output */
#define DG535_DO_CD                     7       /* CD and -CD output */

#define DG535_OUT_TTL                   0
#define DG535_OUT_NIM                   1
#define DG535_OUT_ECL                   2
#define DG535_OUT_VAR                   3

#define DG535_TRIG_INT                  0
#define DG535_TRIG_EXT                  1
#define DG535_TRIG_SS                   2
#define DG535_TRIG_BUR                  3

/* Error status byte values.
 */
#define DG535_ERR_RECALL                0x40    /* recalled data was corrupt */
#define DG535_ERR_DELAY_RANGE           0x20    /* delay range error */
#define DG535_ERR_DELAY_LINKAGE         0x10    /* delay linkage error */
#define DG535_ERR_CMDMODE               0x08    /* wrong mode for command */
#define DG535_ERR_ERANGE                0x04    /* value out of range */
#define DG535_ERR_NUMPARAM              0x02    /* wrong number of paramters */
#define DG535_ERR_BADCMD                0x01    /* unrecognized command */

/* Instrument status byte values.
 * These bits can generate SRQ if present in the status mask.
 */
#define DG535_STAT_BADMEM               0x40    /* memory contents corrupted */
#define DG535_STAT_SRQ                  0x20    /* service request */
#define DG535_STAT_TOOFAST              0x10    /* trigger rate too high */
#define DG535_STAT_PLL                  0x08    /* 80MHz PLL is unlocked */
#define DG535_STAT_TRIG                 0x04    /* trigger has occurred */
#define DG535_STAT_BUSY                 0x02    /* busy with timing cycle */
#define DG535_STAT_CMDERR               0x01    /* command error detected */

static int _interpret_status(gd_t gd, unsigned char status, char *msg);
static int _set_delay(gd_t gd, int out_chan, char *str);

char *prog = "";

static strtab_t out_names[] = {
    { DG535_DO_T0, "t0" },
    { DG535_DO_A,  "a" },
    { DG535_DO_B,  "b" },
    { DG535_DO_AB, "ab" },
    { DG535_DO_C,  "c" },
    { DG535_DO_D,  "d" },
    { DG535_DO_CD, "cd" },
    { 0, NULL },
};
static strtab_t out_modes[] = {
    { DG535_OUT_TTL, "ttl" },
    { DG535_OUT_NIM, "nim" },
    { DG535_OUT_ECL, "ecl" },
    { DG535_OUT_VAR, "var" },
    { 0, NULL },
};
static strtab_t trig_modes[] = {
    { DG535_TRIG_INT, "int" },
    { DG535_TRIG_EXT, "ext" },
    { DG535_TRIG_SS,  "ss" },
    { DG535_TRIG_BUR, "bur" },
    { 0, NULL },
};
static strtab_t impedance[] = {
    { 0, "lo" },
    { 1, "hi" },
    { 0, NULL },
};
static strtab_t slope[] = {
    { 0, "falling" },
    { 1, "rising" },
    { 0, NULL },
};
static strtab_t polarity[] = {
    { 0, "-" },
    { 1, "+" },
    { 0, NULL },
};

static const char *options = OPTIONS_COMMON \
                             "lce:n:dD:qQ:gG:fF:pP:yY:tT:mM:sS:bB:zZ:D:x";
#if HAVE_GETOPT_LONG
#define GETOPT(ac,av,opt,lopt) getopt_long(ac,av,opt,lopt,NULL)
static struct option longopts[] = {
    OPTIONS_COMMON_LONG,
    {"local",           no_argument,       0, 'l'},
    {"clear",           no_argument,       0, 'c'},
    {"display-string",  required_argument, 0, 'e'},
    {"single-shot",     no_argument,       0, 'x'},
    {"out-chan",        required_argument, 0, 'N'},
    {"get-delay",       no_argument,       0, 'd'},
    {"set-delay",       required_argument, 0, 'D'},
    {"get-out-mode",    no_argument,       0, 'q'},
    {"set-out-mode",    required_argument, 0, 'Q'},
    {"get-out-ampl",    no_argument,       0, 'g'},
    {"set-out-ampl",    required_argument, 0, 'G'},
    {"get-out-offset",  no_argument,       0, 'f'},
    {"set-out-offset",  required_argument, 0, 'F'},
    {"get-out-polarity",no_argument,       0, 'p'},
    {"set-out-polarity",required_argument, 0, 'P'},
    {"get-out-z",       no_argument,       0, 'y'},
    {"set-out-z",       required_argument, 0, 'Y'},
    {"get-trig-rate",   no_argument,       0, 't'},
    {"set-trig-rate",   required_argument, 0, 'T'},
    {"get-trig-mode",   no_argument      , 0, 'm'},
    {"set-trig-mode",   required_argument, 0, 'M'},
    {"get-trig-slope",  no_argument,       0, 's'},
    {"set-trig-slope",  required_argument, 0, 'S'},
    {"get-burst-count", no_argument,       0, 'b'},
    {"set-burst-count", required_argument, 0, 'B'},
    {"get-trig-z",      no_argument,       0, 'z'},
    {"set-trig-z",      required_argument, 0, 'Z'},
    {0, 0, 0, 0},
};
#else
#define GETOPT(ac,av,opt,lopt) getopt(ac,av,opt)
#endif

static opt_desc_t optdesc[] = {
    OPTIONS_COMMON_DESC,
    { "l",  "local",          "return instrument to front panel control" },
    { "c",  "clear",          "initialize instrument to default values" },
    { "e",  "display-string", "display string (1-20 chars), empty to clear" },
    { "x",  "single-shot",    "trigger instrument in single shot mode" },
    { "N",  "out-chan",       "select output channel (t0|a|b|ab|c|d|cd)" },
    { "dD", "get/set-delay",  "output delay (another-chan,delay)" },
    { "qQ", "get/set-out-mode", "output mode (ttl|nim|ecl|var)" },
    { "gG", "get/set-out-ampl", "output amplitude (-4:-0.1, +0.1:+4) volts" },
    { "fF", "get/set-out-offset", "output offset (-4:+4) volts" },
    { "pP", "get/set-out-polarity", "output polarity (+|-)" },
    { "yY", "get/set-out-z",  "output impedance (hi|lo)" },
    { "mM", "get/set-trig-mode", "trigger mode (int|ext|ss|bur)" },
    { "tT", "get/set-trig-rate", "trigger rate (0.001hz:1.000mhz)" },
    { "sS", "get/set-trig-slope", "trigger slope (rising|falling)" },
    { "bB", "get/set-burst-count", "trigger burst count (2:19)" },
    { "zZ", "get/set-trig-z", "trigger input impedance (hi|lo)" },
    { 0, 0, 0 },
};

int
main(int argc, char *argv[])
{
    int c, out_chan = -1;
    int tmpi, exit_val = 0;
    int print_usage = 0, out_chan_needed = 0;
    double tmpd;
    char tmpstr[1024];
    gd_t gd;

    gd = gpib_init_args(argc, argv, options, longopts, INSTRUMENT,
                        _interpret_status, 0, &print_usage);
    if (print_usage)
        usage(optdesc);
    if (!gd)
        exit(1);

    /* preparse args to get output channel */
    while ((c = GETOPT(argc, argv, options, longopts)) != EOF) {
        switch (c) {
            case 'N': /* --out-chan */
                out_chan = rfindstr(out_names, optarg);
                if (out_chan == -1) {
                    fprintf(stderr, "%s: bad --out-chan argument\n", prog);
                    exit_val = 1;
                    goto done;
                }
                break;
            case 'q':
            case 'Q':
            case 'g':
            case 'G':
            case 'f':
            case 'F':
            case 'p':
            case 'P':
            case 'y':
            case 'Y':
            case 'd':
            case 'D':
                out_chan_needed++;
                break;
        }
    }
    if (out_chan == -1 && out_chan_needed) {
        fprintf(stderr, "%s: one or more arguments require --out-chan\n", prog);
        exit_val = 1;
        goto done;
    }
    if (out_chan != -1 && !out_chan_needed) {
        fprintf(stderr, "%s: no arguments require --out-chan\n", prog);
        exit_val = 1;
        goto done;
    }

    optind = 0;
    while ((c = GETOPT(argc, argv, options, longopts)) != EOF) {
        switch (c) {
            OPTIONS_COMMON_SWITCH
            case 'o':
                break;
            case 'l': /* --local */
                gpib_loc(gd);
                break;
            case 'c': /* --clear */
                gpib_clr(gd, 0);
                gpib_wrtf(gd, "CL");
                break;
            case 'e': /* --display-string */
                gpib_wrtf(gd, "DS %s", optarg);
                break;
            case 'x': /* --single-shot */
                gpib_wrtf(gd, "SS");
                break;
            case 'd': /* --get-delay */
                gpib_wrtf(gd, "DT %d", out_chan);
                if (gpib_rdf(gd, "%d,%lf", &tmpi, &tmpd) != 2) {
                    fprintf(stderr, "%s: error parsing DT response\n", prog);
                    exit_val = 1;
                    break;
                }
                printf("delay: %s + %lfs\n", findstr(out_names, tmpi), tmpd);
                break;
            case 'D': /* --set-delay */
                if (!_set_delay(gd, out_chan, optarg))
                    exit_val = 1;
                break;
            case 'q' : /* --get-out-mode */
                gpib_wrtf(gd, "OM %d", out_chan);
                if (gpib_rdf(gd, "%d", &tmpi) != 1) {
                    fprintf(stderr, "%s: error parsing OM response\n", prog);
                    exit_val = 1;
                    break;
                }
                printf("output mode: %s\n", findstr(out_modes, tmpi));
                break;
            case 'Q': /* --set-out-mode */
                gpib_wrtf(gd, "OM %d,%d", out_chan, rfindstr(out_modes, optarg));
                break;
            case 'g': /* --get-out-ampl */
                gpib_wrtf(gd, "OA %d", out_chan);
                gpib_rdstr(gd, tmpstr, sizeof(tmpstr));
                printf("output amplitude: %s\n", tmpstr);
                break;
            case 'G': /* --set-out-ampl */
                gpib_wrtf(gd, "OA %d,%s", out_chan, optarg);
                break;
            case 'f' : /* --get-out-offset */
                gpib_wrtf(gd, "OO %d", out_chan);
                gpib_rdstr(gd, tmpstr, sizeof(tmpstr));
                printf("output offset: %s\n", tmpstr);
                break;
            case 'F': /* --set-out-offset */
                gpib_wrtf(gd, "OO %d,%s", out_chan, optarg);
                break;
            case 'p': /* --get-out-polarity */
                gpib_wrtf(gd, "OP %d", out_chan);
                if (gpib_rdf(gd, "%d", &tmpi) != 1) {
                    fprintf(stderr, "%s: error parsing OP response\n", prog);
                    exit_val = 1;
                    break;
                }
                printf("output polarity: %s\n", findstr(polarity, tmpi));
                break;
            case 'P': /* --set-out-polarity */
                gpib_wrtf(gd, "OP %d,%d", out_chan, rfindstr(polarity, optarg));
                break;
            case 'y': /* --get-out-z */
                gpib_wrtf(gd, "TZ %d", out_chan);
                if (gpib_rdf(gd, "%d", &tmpi) != 1) {
                    fprintf(stderr, "%s: error parsing TZ response\n", prog);
                    exit_val = 1;
                    break;
                }
                printf("output impedance: %s\n", findstr(impedance, tmpi));
                break;
            case 'Y': /* --set-out-z */
                gpib_wrtf(gd, "TZ %d,%d", out_chan,rfindstr(impedance, optarg));
                break;
            case 'm' : /* --get-trig-mode */
                gpib_wrtf(gd, "TM");
                if (gpib_rdf(gd, "%d", &tmpi) != 1) {
                    fprintf(stderr, "%s: error parsing TM response\n", prog);
                    exit_val = 1;
                    break;
                }
                printf("trigger mode: %s\n", findstr(trig_modes, tmpi));
                break;
            case 'M' : /* --set-trig-mode */
                gpib_wrtf(gd, "TM %d", rfindstr(trig_modes, optarg));
                break;
            case 't': /* --get-trig-rate */
                gpib_wrtf(gd, "TM");
                if (gpib_rdf(gd, "%d", &tmpi) != 1) {
                    fprintf(stderr, "%s: error parsing TM response\n", prog);
                    exit_val = 1;
                    break;
                }
                gpib_wrtf(gd, "TR %d", tmpi == DG535_TRIG_INT ? 0 : 1);
                if (gpib_rdf(gd, "%lf", &tmpd) != 1) {
                    fprintf(stderr, "%s: error parsing TR response\n", prog);
                    exit_val = 1;
                    break;
                }
                printf("trigger rate: %lfhz\n", tmpd);
                break;
            case 'T': /* --set-trig-rate */
                if (freqstr(optarg, &tmpd) < 0) {
                    fprintf(stderr, "%s: error parsing trigger rate\n", prog);
                    fprintf(stderr, "%s: use units of: %s\n", prog, FREQ_UNITS);
                    exit_val = 1;
                    break;
                }
                gpib_wrtf(gd, "TM");
                if (gpib_rdf(gd, "%d", &tmpi) != 1) {
                    fprintf(stderr, "%s: error parsing TM response\n", prog);
                    exit_val = 1;
                    break;
                }
                gpib_wrtf(gd, "TR %d,%lf", tmpi==DG535_TRIG_INT ? 0 : 1, tmpd);
                break;
            case 's' : /* --get-trig-slope */
                gpib_wrtf(gd, "TS", out_chan);
                if (gpib_rdf(gd, "%d", &tmpi) != 1) {
                    fprintf(stderr, "%s: error parsing TZ response\n", prog);
                    exit_val = 1;
                } else
                    printf("trigger slope: %s\n", findstr(slope, tmpi));
                break;
            case 'S' : /* --set-trig-slope */
                gpib_wrtf(gd, "TS %d", rfindstr(slope, optarg));
                break;
            case 'b' : /* --get-trig-count */
                gpib_wrtf(gd, "BC");
                gpib_rdstr(gd, tmpstr, sizeof(tmpstr));
                printf("burst count: %s\n", tmpstr);
                break;
            case 'B' : /* --set-trig-count */
                gpib_wrtf(gd, "BC %s", optarg);
                break;
            case 'z' : /* --get-trig-z */
                gpib_wrtf(gd, "TZ 0", out_chan);
                if (gpib_rdf(gd, "%d", &tmpi) != 1) {
                    fprintf(stderr, "%s: error parsing TZ response\n", prog);
                    exit_val = 1;
                    break;
                }
                printf("trigger impedance: %s\n", findstr(impedance, tmpi));
                break;
            case 'Z' : /* --set-trig-z */
                gpib_wrtf(gd, "TZ 0,%d", rfindstr(impedance, optarg));
                break;
        }
    }

done:
    gpib_fini(gd);
    exit(exit_val);
}

static void
_check_err(gd_t gd)
{
    char buf[16];
    int status;

    /* check error status */
    gpib_wrtf(gd, "ES");
    gpib_rdstr(gd, buf, sizeof(buf));
    if (*buf < '0' || *buf > '9') {
        fprintf(stderr, "%s: error reading error status byte\n", prog);
        exit(1);
    }
    status = strtoul(buf, NULL, 10);
    if (status & DG535_ERR_RECALL) 
        fprintf(stderr, "%s: recalled data was corrupt\n", prog);
    if (status & DG535_ERR_DELAY_RANGE)
        fprintf(stderr, "%s: delay range error\n", prog);
    if (status & DG535_ERR_DELAY_LINKAGE)
        fprintf(stderr, "%s: delay linkage error\n", prog);
    if (status & DG535_ERR_CMDMODE)
        fprintf(stderr, "%s: wrong mode for command\n", prog);
    if (status & DG535_ERR_ERANGE)
        fprintf(stderr, "%s: value out of range\n", prog);
    if (status & DG535_ERR_NUMPARAM)
        fprintf(stderr, "%s: wrong number of parameters\n", prog);
    if (status & DG535_ERR_BADCMD)
        fprintf(stderr, "%s: unrecognized command\n", prog);
}

static int
_interpret_status(gd_t gd, unsigned char status, char *msg)
{
    int err = 0;

    if (status & DG535_STAT_BADMEM) {
        fprintf(stderr, "%s: memory contents corrupted\n", prog);
        err = 1;
    }
    if (status & DG535_STAT_SRQ) {
        fprintf(stderr, "%s: SRQ pending\n", prog);
        err = 1;
    }
    if (status & DG535_STAT_TOOFAST) {
        fprintf(stderr, "%s: trigger rate too high\n", prog);
        err = 1;
    }
    if (status & DG535_STAT_PLL) {
        fprintf(stderr, "%s: 80MHz PLL is unlocked\n", prog);
        err = 1;
    }
    /*if (status & DG535_STAT_TRIG)
        fprintf(stderr, "%s: trigger has occurred\n", prog);*/
    /*if (status & DG535_STAT_BUSY)
        fprintf(stderr, "%s: busy with timing cycle\n", prog);*/
    if (status & DG535_STAT_CMDERR) {
        _check_err(gd);
        err = 1;
    }
    return err;
}

static int
_set_delay(gd_t gd, int out_chan, char *str)
{
    double f;
    char *cstr, *dstr;
    int follow_chan;
    double delay;

    if (!(cstr = strtok(str, ",")) || !(dstr = strtok(NULL, ","))) {
        fprintf(stderr, "%s: error parsing delay string\n", prog);
        return 0;
    }
    if ((follow_chan = rfindstr(out_names, cstr)) == -1) {
        fprintf(stderr, "%s: use a legal output (t0|a|b|ab|c|d|cd)\n", prog);
        return 0;
    }
    if (freqstr(dstr, &f) == -1) {
        fprintf(stderr, "%s: specify time units in s, ms, us, ns, ps\n", prog);
        return 0;
    }
    delay = 1.0/f;
    gpib_wrtf(gd, "DT %d,%d,%lf", out_chan, follow_chan, delay);
    return 1;
}

/*
 * vi:tabstop=4 shiftwidth=4 expandtab
 */
